src/input_common/joycon/jc_adapter.cpp

```diff
@@ -0,0 +1,372 @@
+    }
+}
+
+void Joycons::Set_rumble(f32 amp_high, f32 amp_low, f32 freq_high, f32 freq_low) {
+    hdRumble.freq_high = std::clamp(freq_high, 81.75177f, 1252.572266f);
+    hdRumble.freq_low = std::clamp(freq_low, 40.875885f, 626.286133f);
+    hdRumble.amp_high = std::clamp(amp_high, 0.0f, 1.0f);
+    hdRumble.amp_low = std::clamp(amp_low, 0.0f, 1.0f);
+    SendRumble();
+}
+void Joycons::SendRumble() {
+    u8* buffer = new u8[49];
+    for (int i = 0; i < 49; ++i) {
+        buffer[i] = 0;
+    }
+    buffer[0] = 0x10;
+    buffer[1] = get_counter();
+    if (hdRumble.amp_high == 0) {
+        buffer[2] = 0x00;
+        buffer[3] = 0x01;
+        buffer[4] = 0x40;
+        buffer[5] = 0x40;
+    } else {
+        u16 encoded_hf = static_cast<u16>(roundf(128 * log2f(hdRumble.freq_high * 0.1f)) - 0x180);
+        u8 encoded_lf = static_cast<u8>(roundf(32 * log2f(hdRumble.freq_low * 0.1f)) - 0x40);
+        u8 encoded_hamp = 0;
+
+        if (hdRumble.amp_high < 0.01182) {
+            encoded_hamp = static_cast<u8>(roundf(pow(hdRumble.amp_high, 1.7f) * 7561));
+        } else if (hdRumble.amp_high < 0.11249) {
+            encoded_hamp = static_cast<u8>(roundf((log(hdRumble.amp_high) * 11.556f) + 55.3f));
+        } else if (hdRumble.amp_high < 0.22498) {
+            encoded_hamp = static_cast<u8>(roundf((log2f(hdRumble.amp_high) * 32) + 131));
+        } else {
+            encoded_hamp = static_cast<u8>(roundf((log2f(hdRumble.amp_high) * 64) + 200));
+        }
+
+        // encoded_lamp should use amp_low data, for now sticking to documentation
+        u8 encoded_lamp = encoded_hamp >> 1;
+        u8 parity = (encoded_lamp % 2) * 0x80;
+        encoded_lamp >>= 1;
+        encoded_lamp += 0x40;
+
+        buffer[2] = encoded_hf & 0xff;
+        buffer[3] = (encoded_hf >> 8) & 0xff;
+        buffer[4] = encoded_lf & 0xff;
+
+        buffer[3] |= encoded_hamp;
+        buffer[4] |= parity;
+        buffer[5] = encoded_lamp;
+    }
+
+    // Have the same data for left and right
+    for (int i = 0; i < 4; ++i) {
+        buffer[6 + i] = buffer[i + 2];
+    }
+    Send_Write(buffer, 49);
+}
+
+void Joycons::Send_Write(u8* buffer, int size) {
+    /*for (int i = 0; i < report_len; ++i) {
+        printf("%02x ", buffer[i]);
+    }
+    printf("\n");*/
+    hid_write(joycon_handle, buffer, size);
+}
+
+u8* Joycons::SubCommand(u8 sc, u8* buffer, int size) {
+    u8* buffer_ = new u8[report_len];
+    for (int i = 0; i < report_len; ++i) {
+        buffer_[i] = 0;
+    }
+
+    buffer_[0] = 0x1;
+    buffer_[1] = get_counter();
+    for (int i = 0; i < 8; i++) {
+        buffer_[2 + i] = default_buffer[i];
+    }
+    buffer_[10] = sc;
+    for (int i = 0; i < size; i++) {
+        buffer_[11 + i] = buffer[i];
+    }
+
+    Send_Write(buffer_, size + 11);
+    return GetResponse(sc);
+}
+
+u8* Joycons::GetResponse(u8 sc) {
+    int tries = 0;
+    u8* buffer = new u8[report_len];
```

Comment:
```suggestion
    std::vector<u8> buffer(report_len);
```

The for loop below this can then be removed.